File: Models/Comparison.m
#import "CIAppDelegate.h"
#import "Comparison.h"
#import "Stock.h"
#import "sqlite3.h"

@implementation Comparison

- (instancetype) init {
    self = [super init];
    [self setStockList:[NSMutableArray arrayWithCapacity:4]];
    return self;
}

+ (NSMutableArray *) listAll:(NSString *)myDbPath {
        
    typedef NS_ENUM(NSInteger, ListAllColumnIndex) {
        ListAllColumnComparisonId,
        ListAllColumnComparisonStockId,
        ListAllColumnStockId,
        ListAllColumnSymbol,
        ListAllColumnStartDate,
        ListAllColumnHasFundamentals,
        ListAllColumnChartType,
        ListAllColumnDaysAgo,
        ListAllColumnColor,
        ListAllColumnFundamentalList,
        ListAllColumnTechnicalList
    };
    sqlite3 *db;
    
    if (sqlite3_open_v2([myDbPath UTF8String], &db, SQLITE_OPEN_READONLY | SQLITE_OPEN_NOMUTEX, NULL) != SQLITE_OK) {
        return [NSMutableArray array];  // To avoid optionals, return an empty array instead of nil
    }
        
	sqlite3_stmt *statement;
    
    if (sqlite3_prepare_v2(db, "SELECT K.rowid, CS.rowid, stockId, symbol, startDate, hasFundamentals, chartType, daysAgo, color, fundamentals, technicals FROM comparison K JOIN comparisonStock CS on K.rowid = CS.comparisonId JOIN stock ON stock.rowid = stockId ORDER BY K.rowid, CS.rowId", -1, &statement, NULL) != SQLITE_OK)
    {
        DLog(@"new SQL failed");
    }
	
    NSMutableArray *list = [NSMutableArray arrayWithCapacity:25];	
    Stock *stock;
    Comparison *comparison;
    NSInteger lastComparisonId = 0;
    NSString *title = @"";
    
    while(sqlite3_step(statement) == SQLITE_ROW) {
        
        if (lastComparisonId != sqlite3_column_int(statement, 0)) {
            comparison = [[self alloc] init];
            [comparison setId: sqlite3_column_int(statement, 0)];
           // DLog(@"allocating new comparison with id %d", [comparison id]);
            lastComparisonId = [comparison id];
            [list addObject:comparison];
            [comparison release];
            title = @"";
        } else {
           // DLog(@"allocating existing series id %d", lastComparisonId);
        }
        
        stock = [[Stock alloc] init];
        stock.comparisonStockId = sqlite3_column_int(statement, ListAllColumnComparisonStockId);
        stock.id = sqlite3_column_int(statement, ListAllColumnStockId);
                              
        [stock setSymbol:[NSString stringWithUTF8String:(const char *) sqlite3_column_text(statement, ListAllColumnSymbol)]];
        
       // DLog(@"symbol is %@", [stock symbol]);
        
        title = [title stringByAppendingFormat:@"%@ ", stock.symbol];

        stock.startDateString = [NSString stringWithUTF8String:(const char *) sqlite3_column_text(statement, ListAllColumnStartDate)];
        // startDateString will be converted to NSDate by [StockData init] as price data is loaded
        
        stock.hasFundamentals = sqlite3_column_int(statement, ListAllColumnHasFundamentals);
        
        stock.chartType = sqlite3_column_int(statement, ListAllColumnChartType);
        stock.daysAgo = sqlite3_column_int(statement, ListAllColumnDaysAgo);
        
        if (sqlite3_column_bytes(statement, ListAllColumnColor) > 2) {
            [stock setColorWithHexString:[NSString stringWithUTF8String:(const char *)sqlite3_column_text(statement, ListAllColumnColor)]];
        } else {
            [stock setColorWithHexString:@"009900"];   // green (and by convention, red)
        }
                
        if (sqlite3_column_bytes(statement, ListAllColumnFundamentalList) > 2) {
            const char *fundamentals = (const char *)sqlite3_column_text(statement, ListAllColumnFundamentalList);
            [stock setFundamentalList:[NSString stringWithUTF8String:fundamentals]];
        }
        
        if (sqlite3_column_bytes(statement, ListAllColumnTechnicalList) > 2) {
            const char *technicals = (const char *)sqlite3_column_text(statement, ListAllColumnTechnicalList);
            [stock setTechnicalList:[NSString stringWithUTF8String:technicals]];
        }
        
        [comparison.stockList addObject:stock];
        [comparison setTitle:title];
        [stock release];

    }
    sqlite3_finalize(statement);	
    sqlite3_close(db);
    
    return list;
}

- (void) saveToDb {
    
    sqlite3 *db;
    sqlite3_stmt *statement;
    
    if (sqlite3_open_v2([[NSString stringWithFormat:@"%@/Documents/charts.db", NSHomeDirectory()] UTF8String], &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_NOMUTEX, NULL) != SQLITE_OK) {
        return;
    }
    
    if (self.id == 0) {
       // DLog(@"comparisonId is %d so insertSQL to get an id for each comparisonStock", id);
        sqlite3_prepare_v2(db, "INSERT INTO comparison (sort) VALUES (0)", -1, &statement, NULL);
        sqlite3_step(statement);
        sqlite3_finalize(statement);
        NSInteger insertedRowid = (NSInteger) sqlite3_last_insert_rowid(db);    // cast to NSInteger since we don't support billions of comparisons...
        [self setId:insertedRowid];
    }
        
    for (NSInteger i = 0; i < self.stockList.count; i++) {
        Stock *stock = [self.stockList objectAtIndex:i];

        if (stock.comparisonStockId > 0) {
            sqlite3_prepare(db, "UPDATE comparisonStock SET daysAgo = ?, chartType = ?, color = ?, fundamentals = ?, technicals = ? WHERE rowid = ?", -1, &statement, NULL);
   
            sqlite3_bind_int64(statement, 1, stock.daysAgo);
            sqlite3_bind_int64(statement, 2, stock.chartType);
            sqlite3_bind_text(statement, 3, [[stock hexFromColor] UTF8String], 6, SQLITE_STATIC); // STATIC = don't free space
            sqlite3_bind_text(statement, 4, [[stock fundamentalList] UTF8String], (int)[stock fundamentalList].length, SQLITE_STATIC);
            sqlite3_bind_text(statement, 5, [[stock technicalList] UTF8String], (int)[stock technicalList].length, SQLITE_STATIC);
            sqlite3_bind_int64(statement, 6, stock.comparisonStockId);

            if(sqlite3_step(statement)==SQLITE_DONE){
          //      // DLog(@"updated %d to DB for comparison id %d", stock.comparisonStockId, [self id]);
            } else {
          //      // DLog(@"DB ERROR '%s'.", sqlite3_errmsg(db));
            }
            sqlite3_finalize(statement);
        } else if (stock.comparisonStockId == 0) {
            
            sqlite3_prepare_v2(db, "INSERT OR REPLACE INTO comparisonStock (comparisonId, stockId, daysAgo, chartType, color, fundamentals, technicals) VALUES (?, ?, ?, ?, ?, ?, ?)", -1, &statement, NULL);
            
            sqlite3_bind_int64(statement, 1, [self id]);
            sqlite3_bind_int64(statement, 2, stock.id);
            sqlite3_bind_int64(statement, 3, stock.daysAgo);
            sqlite3_bind_int64(statement, 4, stock.chartType);
            sqlite3_bind_text(statement, 5, [[stock hexFromColor] UTF8String], 6, SQLITE_STATIC); // STATIC = don't free space
            sqlite3_bind_text(statement, 6, [[stock fundamentalList] UTF8String], (int)[stock fundamentalList].length, SQLITE_STATIC);
            sqlite3_bind_text(statement, 7, [[stock technicalList] UTF8String], (int)[stock technicalList].length, SQLITE_STATIC);

            if(sqlite3_step(statement)==SQLITE_DONE){
                stock.comparisonStockId = (NSInteger) sqlite3_last_insert_rowid(db);

             //   DLog(@"inserted into DB as CSID %d for comparison id %d", stock.comparisonStockId, [self id]);
            } else {
              //  DLog(@"DB ERROR '%s'.", sqlite3_errmsg(db));
            }
            sqlite3_finalize(statement);
        }
    }

    sqlite3_close(db);
}


- (void) deleteFromDb {
    
    sqlite3 *db;
    
    if (sqlite3_open_v2([[NSString stringWithFormat:@"%@/Documents/charts.db", NSHomeDirectory()] UTF8String], &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_NOMUTEX, NULL) != SQLITE_OK) {
        return;
    }
    sqlite3_stmt *statement;
    
    if (sqlite3_prepare_v2(db, "DELETE FROM comparisonStock WHERE comparisonId =  ?", -1, &statement, NULL) == SQLITE_OK) {
        
        sqlite3_bind_int64(statement, 1, self.id);
        if(sqlite3_step(statement) != SQLITE_DONE){;
            DLog(@"Delete comparisonStock DB ERROR '%s'.", sqlite3_errmsg(db));
        }
    }
    
    sqlite3_finalize(statement);
    
    if (sqlite3_prepare_v2(db, "DELETE FROM comparison WHERE rowid =  ?", -1, &statement, NULL) == SQLITE_OK) {
        
        sqlite3_bind_int64(statement, 1, self.id);
        if(sqlite3_step(statement) != SQLITE_DONE){;
            DLog(@"Delete Comparison DB ERROR '%s'.", sqlite3_errmsg(db));
        }
    }
    sqlite3_finalize(statement);
    sqlite3_close(db);
}


//- (void) deleteComparisonStockAtIndex:(NSInteger)index {
//
//    if (index >= [self.stockList count]) { // index out of bounds
//        return;
//    }
//
//    Stock *s = [self.stockList objectAtIndex:index];
//
//    [self deleteStock:s];
//}

- (void) deleteStock:(Stock *)s {
    
    sqlite3 *db;
    if (sqlite3_open_v2([[NSString stringWithFormat:@"%@/Documents/charts.db", NSHomeDirectory()] UTF8String], &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_NOMUTEX, NULL) != SQLITE_OK) {
        return;
    }
    
    sqlite3_stmt *statement;
    
    if (sqlite3_prepare_v2(db, "DELETE FROM comparisonStock WHERE stockId=?", -1, &statement, NULL) == SQLITE_OK) {
        
        sqlite3_bind_int64(statement, 1, s.id);    // for comparisonStock table
        
        if(sqlite3_step(statement) == SQLITE_DONE){;
            [self.stockList removeObject:s];
        } else {
            DLog(@"DB ERROR '%s'.", sqlite3_errmsg(db));
        }
    }
    
    sqlite3_finalize(statement);
    sqlite3_close(db);
}

- (NSArray *)chartTypes {
    return [(CIAppDelegate *)[[UIApplication sharedApplication] delegate] chartTypes];
}

- (NSArray *)metrics {
    return [(CIAppDelegate *)[[UIApplication sharedApplication] delegate] metrics];
}

// Count of superset of all keys for all subcharts excluding book value per share, which is an overlay
- (NSArray *) sparklineKeys {

    NSString *keyStrings = @"";
    
    for (Stock *stock in self.stockList) {
        keyStrings = [keyStrings stringByAppendingString:stock.fundamentalList];
    }
 
    keyStrings = [keyStrings stringByReplacingOccurrencesOfString:@"BookValuePerShare," withString:@""];
        
    NSMutableArray *sortedMetrics = [NSMutableArray new];
    
    for (NSArray *category in [self metrics]) {
        for (NSArray *metric in category) {
            NSString *key = [metric objectAtIndex:0];
            if ([keyStrings rangeOfString:key].length > 0) {
                [sortedMetrics addObject:key];
            }
        }
    }
    return sortedMetrics;
}

- (void) resetMinMax {
    [self setMinForKey:[NSMutableDictionary new]];
    [self setMaxForKey:[NSMutableDictionary new]];
}

- (void) updateMinMaxForKey:(nullable NSString *)key withValue:(nullable NSDecimalNumber *)reportValue {
    if (reportValue != nil && [reportValue isEqualToNumber:[NSDecimalNumber notANumber]] == NO) {
        if ([self.minForKey objectForKey:key] == nil || [[self.minForKey objectForKey:key] isEqualToNumber:[NSDecimalNumber notANumber]]) {            
            if ([reportValue compare:[NSDecimalNumber zero]] == NSOrderedAscending) {
                [self.minForKey setObject:reportValue forKey:key];
            } else {
                [self.minForKey setObject:[NSDecimalNumber zero] forKey:key];
            }
           // DLog(@"initializing min to %@ for key %@", [self.minForKey objectForKey:key], key);
            [self.maxForKey setObject:reportValue forKey:key];
        } else {
            if ([reportValue compare:[self.minForKey objectForKey:key]] == NSOrderedAscending) {
              //  DLog(@"report value %@ < %@ self.minForKey", reportValue, [self.minForKey objectForKey:key]);
                [self.minForKey setObject:reportValue forKey:key];
            }
            if ([reportValue compare:[self.maxForKey objectForKey:key]] == NSOrderedDescending) {
             //   DLog(@"report value %@ > %@ maxForkey", reportValue, [maxForKey objectForKey:key]);
                [self.maxForKey setObject:reportValue forKey:key];
            }
        }
    }
}

- (nonnull NSDecimalNumber *) rangeForKey:(nullable NSString *)key {
    if ([self.minForKey objectForKey:key] == nil) {
        return [NSDecimalNumber notANumber];
    }
    return [[self.maxForKey objectForKey:key] decimalNumberBySubtracting:[self.minForKey objectForKey:key]];
}
    
- (nullable NSDecimalNumber *) minForKey:(nullable NSString *)key {
    return [self.minForKey objectForKey:key];
}

- (nullable NSDecimalNumber *) maxForKey:(nullable NSString *)key {
    return [self.maxForKey objectForKey:key];
}



@end

File: ChartInsight/Models/Comparison.swift
//
//  Comparison.swift
//  ChartInsight
//
//  Created by Eric Kennedy on 6/19/23.
//  Copyright © 2023 Chart Insight LLC. All rights reserved.
//

import Foundation

class Comparison: NSObject {
    typealias sqlite3ptr = OpaquePointer?
    let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)
    
    @objc var id: Int = 0
    @objc var stockList: [Stock] = []
    @objc var title: String = ""
    var minMetricValues: [String: NSDecimalNumber] = [:]
    var maxMetricValues: [String: NSDecimalNumber] = [:]
      
    /// Union of all metric keys for stocks in this comparison set
    @objc func sparklineKeys() -> [String] {
        var fundamentalKeys = "";
            
        for stock in stockList {
            fundamentalKeys = fundamentalKeys.appending(stock.fundamentalList)
        }
     
        var sortedMetrics: [String] = []
        
        if let delegate = UIApplication.shared.delegate as? CIAppDelegate {
            for category in delegate.metrics {
                for metric in category {
                    if metric.count > 0 {
                        let metricKey = String(metric[0])
                        if fundamentalKeys.contains(metricKey) {
                            sortedMetrics.append(metricKey)
                        }
                    }
                }
            }
        }
        return sortedMetrics
    }
    
    @objc func resetMinMax() {
        minMetricValues.removeAll(keepingCapacity: true)
        maxMetricValues.removeAll(keepingCapacity: true)
    }
    
    /// Determine min and max values for fundamental metric key
    @objc func updateMinMax(for key: String, value: NSDecimalNumber?) {
        guard value != nil && value != .notANumber else { return }

        if let minValueForKey = minMetricValues[key], minValueForKey != .notANumber {
            if value?.compare(minValueForKey) == .orderedAscending {
                minMetricValues[key] = value
            }
            if let maxValueForKey = maxMetricValues[key],
               value?.compare(maxValueForKey) == .orderedDescending {
                maxMetricValues[key] = value
            }
            
        } else { // minKeyValues[key] == nil or .notANumber
            // Fundamental bar scale should range from zero (or a negative report value) to max
            if (value?.compare(NSDecimalNumber.zero) == .orderedAscending) {
                minMetricValues[key] = value // negative value
            } else {
                minMetricValues[key] = NSDecimalNumber.zero
            }
            maxMetricValues[key] = value
        }
    }
    
    /// Returns notANumber if no values for key
    @objc func range(for key: String) -> NSDecimalNumber {
        if let maxValue = maxMetricValues[key],
           let minValue = minMetricValues[key] {
            return maxValue.subtracting(minValue)
        }
        return NSDecimalNumber.notANumber
    }
    
    @objc func min(for key: String) -> NSDecimalNumber? {
        return minMetricValues[key]
    }
    
    @objc func max(for key: String) -> NSDecimalNumber? {
        return maxMetricValues[key]
    }
    
    static func dbPath() -> String {
        return String(format: "%@/Documents/charts.db", NSHomeDirectory())
    }
    
    @objc func add(_ stock: Stock) {
        stockList.append(stock)
        saveToDb()
    }
    
    /// Insert or update this stock comparison
    @objc func saveToDb() {
        var db: sqlite3ptr = nil
        
        guard SQLITE_OK == sqlite3_open(Comparison.dbPath(), &db) else { return }
        var statement: sqlite3ptr = nil
        var sql = ""
        if id == 0 {
            sql = "INSERT INTO comparison (sort) VALUES (0)"
            guard SQLITE_OK == sqlite3_prepare_v2(db, sql, -1, &statement, nil) else { return }
            sqlite3_step(statement)
            sqlite3_finalize(statement)
            id = Int(sqlite3_last_insert_rowid(db))
        }
        
        for stock in stockList {
            let hexColorUTF8     = NSString(string: stock.hexFromUpColor()).utf8String
            let fundamentalsUTF8 = NSString(string: stock.fundamentalList).utf8String
            let technicalsUTF8   = NSString(string: stock.technicalList).utf8String
            
            if stock.comparisonStockId > 0 {
                sql = "UPDATE comparisonStock SET chartType=?, color=?, fundamentals=?, technicals=? WHERE rowid=?"
                sqlite3_prepare(db, sql, -1, &statement, nil);
    
                sqlite3_bind_int64(statement, 1, Int64(stock.chartType.rawValue));
                sqlite3_bind_text(statement,  2, hexColorUTF8, -1, SQLITE_TRANSIENT)
                sqlite3_bind_text(statement,  3, fundamentalsUTF8, -1, SQLITE_TRANSIENT);
                sqlite3_bind_text(statement,  4, technicalsUTF8, -1, SQLITE_TRANSIENT);
                sqlite3_bind_int64(statement, 5, Int64(stock.comparisonStockId));

                if SQLITE_DONE != sqlite3_step(statement) {
                    print("Save to DB ERROR ", String(cString: sqlite3_errmsg(db)), sql)
                }
                sqlite3_finalize(statement);
            } else if stock.comparisonStockId == 0 {
                sql = "INSERT OR REPLACE INTO comparisonStock (comparisonId, stockId, chartType, color, fundamentals, technicals) VALUES (?, ?, ?, ?, ?, ?)"
                sqlite3_prepare(db, sql, -1, &statement, nil);
                
                sqlite3_bind_int64(statement, 1, Int64(id));
                sqlite3_bind_int64(statement, 2, Int64(stock.id));
                sqlite3_bind_int64(statement, 3, Int64(stock.chartType.rawValue));
                sqlite3_bind_text(statement,  4, hexColorUTF8, -1, SQLITE_TRANSIENT)
                sqlite3_bind_text(statement,  5, fundamentalsUTF8, -1, SQLITE_TRANSIENT);
                sqlite3_bind_text(statement,  6, technicalsUTF8, -1, SQLITE_TRANSIENT);

                if SQLITE_DONE != sqlite3_step(statement) {
                    print("Save to DB ERROR ", String(cString: sqlite3_errmsg(db)), sql)
                }
                sqlite3_finalize(statement);
            }
        }
        sqlite3_close(db);
    }
    
    /// Deletes comparison row and all comparisonStock rows
    @objc func deleteFromDb() {
        var db: sqlite3ptr = nil
        
        guard SQLITE_OK == sqlite3_open(Comparison.dbPath(), &db) else { return }
        var statement: sqlite3ptr = nil
        
        var sql = "DELETE FROM comparisonStock WHERE comparisonId = ?"
        guard SQLITE_OK == sqlite3_prepare_v2(db, sql, -1, &statement, nil) else { return }
        
        sqlite3_bind_int64(statement, 1, Int64(id))
        
        if SQLITE_DONE != sqlite3_step(statement){
            print(String(format: "Delete from comparisonStock DB ERROR '%s'.", sqlite3_errmsg(db)))
        }
        sqlite3_finalize(statement);
        
        sql = "DELETE FROM comparison WHERE rowid = ?"
        guard SQLITE_OK == sqlite3_prepare_v2(db, sql, -1, &statement, nil) else { return }
                
        sqlite3_bind_int64(statement, 1, Int64(id))
        
        if SQLITE_DONE != sqlite3_step(statement){
            print(String(format: "Delete from comparison DB ERROR '%s'.", sqlite3_errmsg(db)))
        }
        sqlite3_finalize(statement);
        sqlite3_close(db);        
    }
    
    /// Delete a single stock from this comparison
    @objc(deleteStock:) func delete(stock: Stock) {
        var db: sqlite3ptr = nil
        
        guard SQLITE_OK == sqlite3_open(Comparison.dbPath(), &db) else { return }
        var statement: sqlite3ptr = nil
        
        let sql = "DELETE FROM comparisonStock WHERE stockId = ?"
        guard SQLITE_OK == sqlite3_prepare_v2(db, sql, -1, &statement, nil) else { return }
        
        sqlite3_bind_int64(statement, 1, Int64(stock.id))
        
        if SQLITE_DONE == sqlite3_step(statement){
            stockList.removeAll(where: {$0 == stock})
        } else {
            print(String(format: "delete(stock:) DB ERROR '%s'.", sqlite3_errmsg(db)))
        }
        sqlite3_finalize(statement);
        sqlite3_close(db);
    }
    
    @objc
    static func listAll() -> [Comparison] {
        var list: [Comparison] = []
        var db: sqlite3ptr = nil
        
        guard SQLITE_OK == sqlite3_open(dbPath(), &db) else { return list }
        var statement: sqlite3ptr = nil
        
        enum CI: Int32 {
            case comparisonId
            case comparisonStockId
            case stockId
            case symbol
            case startDate
            case hasFundamentals
            case chartType
            case color
            case fundamentalList
            case technicalList
        }
        
        let sql = "SELECT K.rowid, CS.rowid, stockId, symbol, startDate, hasFundamentals, chartType, color, fundamentals, technicals FROM comparison K JOIN comparisonStock CS on K.rowid = CS.comparisonId JOIN stock ON stock.rowid = stockId ORDER BY K.rowid, CS.rowId"
       
        guard SQLITE_OK == sqlite3_prepare_v2(db, sql, -1, &statement, nil) else { return list }
        
        var comparison: Comparison? = nil
        var lastComparisonId = 0
        var title = ""
        
        while (SQLITE_ROW == sqlite3_step(statement)) {
            if lastComparisonId != sqlite3_column_int(statement, 0) {
                title = ""
                comparison = Comparison() // only need to create a new comparison after initial one
                lastComparisonId = Int(sqlite3_column_int(statement, 0))
                comparison?.id = lastComparisonId
                list.append(comparison!)
            }
            let stock = Stock()
            stock.comparisonStockId = Int(sqlite3_column_int(statement, CI.comparisonStockId.rawValue))
            stock.id = Int(sqlite3_column_int(statement, CI.stockId.rawValue))
            stock.symbol = String(cString: UnsafePointer(sqlite3_column_text(statement, CI.symbol.rawValue)))
            title = title.appending("\(stock.symbol) ")
            stock.startDateString = String(cString: UnsafePointer(sqlite3_column_text(statement, CI.startDate.rawValue)))
            // startDateString will be converted to NSDate by [StockData init] as price data is loaded
            
            stock.hasFundamentals = 0 < sqlite3_column_int(statement, CI.hasFundamentals.rawValue)
            if let chartType = ChartType(rawValue: Int(sqlite3_column_int(statement, CI.chartType.rawValue))) {
                stock.chartType = chartType
            }
            
            let hexString = String(cString: UnsafePointer(sqlite3_column_text(statement, CI.color.rawValue)))
            
            if hexString != "" {
                stock.setColorWith(hexString: hexString)
            } else {
                stock.setColorWith(hexString: "009900") // upColor = green, down = red
            }
            
            if sqlite3_column_bytes(statement, CI.fundamentalList.rawValue) > 2 {
                stock.fundamentalList = String(cString: UnsafePointer(sqlite3_column_text(statement, CI.fundamentalList.rawValue)))
            }
            
            if sqlite3_column_bytes(statement, CI.technicalList.rawValue) > 2 {
                stock.technicalList = String(cString: UnsafePointer(sqlite3_column_text(statement, CI.technicalList.rawValue)))
            }
            
            comparison?.stockList.append(stock)
            comparison?.title = title
        }
        sqlite3_finalize(statement)
        sqlite3_close(db)
            
        return list
    }
    
}
